Modules/_ssl.c

```diff
@@ -3289,6 +3291,200 @@ _ssl__SSLContext_load_cert_chain_impl(PySSLContext *self, PyObject *certfile,
+    ret = SSL_CTX_use_certificate(ctx, usecert);
+    if (ERR_peek_error() != 0) {
+        ret = 0;
+        goto end;
+    }
+
+    /* Certificate is set up, proceed reading chain certificates. */
+    r = SSL_CTX_clear_chain_certs(ctx);
+    if (r == 0) {
+        ret = 0;
+        goto end;
+    }
+
+    while (1) {
+        cacert = PEM_read_bio_X509(bio, NULL, pwcb, pwcb_data);
+
+        if (cacert == NULL)
+            /* Error: BIO EOF or something critical */
+            break;
+
+        r = SSL_CTX_add0_chain_cert(ctx, cacert);
+        if (!r) {
+            X509_free(cacert);
+            ret = 0;
+            goto end;
+        }
+    }
+
+    err = ERR_peek_last_error();
+    if (ERR_GET_LIB(err) == ERR_LIB_PEM
+        && ERR_GET_REASON(err) == PEM_R_NO_START_LINE)
+        /* BIO EOF, expected. */
+        ERR_clear_error();
+    else
+        /* A critical error. */
+        ret = 0;
+
+ end:
+    X509_free(usecert);
+    return ret;
+}
+
+
+/*[clinic input]
+_ssl._SSLContext._load_cert_chain_pem_from_bio
+    certbio: object(subclass_of="&PySSLMemoryBIO_Type", type="PySSLMemoryBIO *")
+    keybio: object(subclass_of="&PySSLMemoryBIO_Type", type="PySSLMemoryBIO *")
+    password: object = NULL
+
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLContext__load_cert_chain_pem_from_bio_impl(PySSLContext *self,
+                                                    PySSLMemoryBIO *certbio,
+                                                    PySSLMemoryBIO *keybio,
+                                                    PyObject *password)
+/*[clinic end generated code: output=1fa9f45cbdeddf1c input=4ba4aff7f210d0fd]*/
+{
+    /*
+    Note that _password_callback() starts by invoking
+
+        PySSL_END_ALLOW_THREADS_S(pw_info->thread_state);
+
+    end finishes with invoking
+
+        PySSL_BEGIN_ALLOW_THREADS_S(pw_info->thread_state);
+
+    and all OpenSSL API calls that potentially call the password callback are
+    required to be wrapped in
+
+        PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
+        [...]
+        PySSL_END_ALLOW_THREADS_S(pw_info.thread_state);
+    */
+
+    EVP_PKEY *private_key = NULL;
+    pem_password_cb *orig_passwd_cb = self->ctx->default_passwd_callback;
+
+    void *orig_passwd_userdata = self->ctx->default_passwd_callback_userdata;
+    _PySSLPasswordInfo pw_info = { NULL, NULL, NULL, 0, 0 };
+    int r;
+
+    errno = 0;
+    ERR_clear_error();
+
+    if (password && password != Py_None) {
+        if (PyCallable_Check(password)) {
+            pw_info.callable = password;
+        } else if (!_pwinfo_set(&pw_info, password,
+                                "password should be a string or callable")) {
+            goto error;
+        }
+        SSL_CTX_set_default_passwd_cb(self->ctx, _password_callback);
+        SSL_CTX_set_default_passwd_cb_userdata(self->ctx, &pw_info);
+    }
+
+
+    PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
```

Comment:
Can a password be asked for the certificates (as opposed to the private key)? This sounds a bit weird.

Reply:
Hey @pitrou. I got around to looking into this again. My argument above was not the correct argument to answer your question. I went into the code and docs again and understood the reasoning again. There are two points of relevance here:

**1) cert files might indeed be encrypted**

From https://www.openssl.org/docs/man1.0.2/ssl/SSL_CTX_use_certificate_chain_file.html:

> The private keys loaded from file can be encrypted. In order to successfully load encrypted keys, a function returning the passphrase must have been supplied, see SSL_CTX_set_default_passwd_cb. (**Certificate files might be encrypted as well** from the technical point of view, it however does not make sense as the data in the certificate is considered public anyway.)

(emphasis mine)

**2) Consistency**

That is the code block in question from my patch:

```
    PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
    r = _openssl_use_certificate_chain_from_bio(self->ctx, certbio->bio);
    PySSL_END_ALLOW_THREADS_S(pw_info.thread_state);
```

whereas ` _openssl_use_certificate_chain_from_bio()` is by definition meant to _closely_ resemble OpenSSL's `SSL_CTX_use_certificate_chain_file()`.  Both *can* handle the password callback. As of reason (1), I suppose. From the [source](https://github.com/openssl/openssl/blob/d72a00416a0691bfd4920008767221bb4082a2ed/ssl/ssl_rsa.c#L589) of `SSL_CTX_use_certificate_chain_file()` in OpenSSL 1.1.0:
```
    x = PEM_read_bio_X509_AUX(in, NULL, passwd_callback,
                              passwd_callback_userdata);
```

 *If* `_openssl_use_certificate_chain_from_bio()` or `SSL_CTX_use_certificate_chain_file()` triggers the password callback I believe we better have the wrapper in place.

That's probably what one of the earlier authors thought, too, and so the same ALLOW_THREADS wrapper around `SSL_CTX_use_certificate_chain_file()` has always been [present in CPython so far](https://github.com/python/cpython/blob/40c54d5e1aaab91cb7df71f735112d20b5e5b755/Modules/_ssl.c#L3323):
```
    PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
    r = SSL_CTX_use_certificate_chain_file(self->ctx,
        PyBytes_AS_STRING(certfile_bytes));
    PySSL_END_ALLOW_THREADS_S(pw_info.thread_state);
```

Notably, I do not think that we cover the case of loading an encrypted certificate (where there is no key also present in the cert file) within `test_ssl`.

Reply:
I have added a commit with a hopefully clarifying code comment.